### PR TITLE
Infura calls: Only request data for cards we show

### PR DESCRIPTION
**Pull request by sennett**
# [Only request data for cards we show](https://app.shortcut.com/oazo-apps/story/5171/don-t-request-all-product-card-data-when-only-showing-a-few-product-cards)

Infura calls reduction:

- Load homepage: 324 to about 40 calls.
- Navigate around website for a 10-block user-session: 
  - before (10 blocks duration): 4000 calls
  - after (7 blocks duration): 347 calls

## Changes 👷‍♀️
- Only request data for product cards that are displayed.   Home page, assets pages, product pages.
  
## How to test 🧪
  <Please explain how to test your changes>
- All cards display across all pages.  Verify button action.
- Open/close a few positions.

Some images
-----

Before:

Pasted Graphic 7

After:

Pasted Graphic 6

### CODE REVIEW

File: components/ProductCardsFilter.tsx [hunk 1]
```diff
@@ -1,26 +1,54 @@
...
   mapUrlFragmentToFilter,
+  ProductCardData,
   ProductLandingPagesFilter,
   ProductLandingPagesFiltersKeys,
 } from '../helpers/productCards'
+import { useAppContext } from './AppContextProvider'
 import { ProductCardsSelect } from './ProductCardsSelect'
+import { ProductCardsWrapper } from './ProductCardsWrapper'
 
 interface TokenTabsProps {
```
Comment (jakubswierczek) [inline components/ProductCardsFilter.tsx:20]: I know that it's not your PR, but can you rename it to something more relevant to the component?

File: components/ProductCardsFilter.tsx [hunk 2]
```diff
@@ -1,26 +1,54 @@
...
   filters: Array<ProductLandingPagesFilter>
-  children: (token: ProductLandingPagesFiltersKeys) => ReactNode
+  productCardComponent: (props: { cardData: ProductCardData }) => JSX.Element
   selectedFilter?: string
+  filterCardsFunction: ({
```
Comment (jakubswierczek) [inline components/ProductCardsFilter.tsx:24]: introduce a type for that function signature.

File: components/ProductCardsFilter.tsx [hunk 3]
```diff
@@ -1,26 +1,54 @@
...
+    ilkToTokenMapping: ilkToEntryToken,
+    cardsFilter: currentFilter,
+  })
+  const { productCardsData$ } = useAppContext()
+  const [productCardsData, productCardsDataError] = useObservable(
+    productCardsData$(productsToDisplay.map(({ ilk }) => ilk)),
+  )
   const [hover, setHover] = useState('')
-
+  const ProductCardComponent = productCardComponent
```
Comment (jakubswierczek) [inline components/ProductCardsFilter.tsx:51]: I think this is unnecessary

File: components/ProductCardsFilter.tsx [hunk 4]
```diff
@@ -78,8 +106,24 @@ export function ProductCardsFilter({ filters, children, selectedFilter }: TokenT
           currentFilter={currentFilter}
         />
       </Flex>
-
-      {children(currentFilter)}
+      <WithErrorHandler error={[productCardsDataError]}>
+        <WithLoadingIndicator
+          value={[productCardsData]}
+          customLoader={
+            <Flex sx={{ alignItems: 'flex-start', justifyContent: 'center', height: '500px' }}>
```
Comment (jakubswierczek) [inline components/ProductCardsFilter.tsx:113]: move `customLoader` to new component

File: features/asset/AssetView.tsx [hunk 5]
```diff
@@ -25,73 +25,69 @@ function Loader() {
...
-  type: ProductTypes
+type TabContentProps = {
   renderProductCard: (props: { cardData: ProductCardData }) => JSX.Element
   ilks: string[]
-  productCardsData: ProductCardData[]
-}) {
+}
+
+function TabContent(props: TabContentProps) {
   const ProductCard = props.renderProductCard
```
Comment (jakubswierczek) [inline features/asset/AssetView.tsx:34]: what is the point of creating a new variable here?

File: features/homepage/HomepageView.tsx [hunk 6]
```diff
@@ -55,15 +56,42 @@ function TabContent(props: {
...
+            </ProductCardsWrapper>
+          )}
+        </WithLoadingIndicator>
+      </WithErrorHandler>
     </Flex>
   )
 }
 
+// we need these wrappers to avoid react trying to render the wrong card types for the wrong ilks
+function TabContentMultiply(props: TabContentProps) {
```
Comment (jakubswierczek) [inline features/homepage/HomepageView.tsx:83]: I saw that before...  Can we merge it with `AssetView` ?

File: components/ProductCardsFilter.tsx (hunk 3 above)
Comment (sennett) [reply components/ProductCardsFilter.tsx:51]: Ah no React needs these to be caps: https://stackoverflow.com/a/49273625/614523

File: features/asset/AssetView.tsx (hunk 5 above)
Comment (sennett) [reply features/asset/AssetView.tsx:34]: Same as above - react needs these to have caps. The other option (I think) would be to have the prop be a capital letter.

Comment (sennett) [reply features/asset/AssetView.tsx]: See above - react needs caps names.

File: features/homepage/HomepageView.tsx (hunk 6 above)
Comment (sennett) [reply features/homepage/HomepageView.tsx]: Done.  Moved product cards to their own folder and made a container.

**Review by jakubswierczek — APPROVED** [review]

**Pull request merged**